src/System.Linq/src/System/Linq/AppendPrepend.cs

```diff
@@ -0,0 +1,715 @@
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        public static IEnumerable<TSource> Append<TSource>(this IEnumerable<TSource> source, TSource element)
+        {
+            if (source == null)
+            {
+                throw Error.ArgumentNull(nameof(source));
+            }
+
+            AppendPrependIterator<TSource> ap = source as AppendPrependIterator<TSource>;
+            if (ap != null)
+            {
+                return ap.Append(element);
+            }
+
+            return new Append1Iterator<TSource>(source, element);
+        }
+
+        public static IEnumerable<TSource> Prepend<TSource>(this IEnumerable<TSource> source, TSource element)
+        {
+            if (source == null)
+            {
+                throw Error.ArgumentNull(nameof(source));
+            }
+
+            AppendPrependIterator<TSource> ap = source as AppendPrependIterator<TSource>;
+            if (ap != null)
+            {
+                return ap.Prepend(element);
+            }
+
+            return new Prepend1Iterator<TSource>(source, element);
+        }
+
+        private abstract class AppendPrependIterator<TSource> : Iterator<TSource>, IIListProvider<TSource>
+        {
+            protected readonly IEnumerable<TSource> _source;
+            private IEnumerator<TSource> _enumerator;
+
+            protected AppendPrependIterator(IEnumerable<TSource> source)
+            {
+                Debug.Assert(source != null);
+                _source = source;
+            }
+
+            protected void GetSourceEnumerator()
+            {
+                Debug.Assert(_enumerator == null);
+                _enumerator = _source.GetEnumerator();
+            }
+
+            protected bool LoadFromEnumerator()
+            {
+                if (_enumerator.MoveNext())
+                {
+                    _current = _enumerator.Current;
+                    return true;
+                }
+
+                DisposeEnumerator();
+                return false;
+            }
+
+            private void DisposeEnumerator()
+            {
+                if (_enumerator != null)
+                {
+                    _enumerator.Dispose();
+                    _enumerator = null;
+                }
+            }
+
+            public sealed override void Dispose()
+            {
+                DisposeEnumerator();
+
+                base.Dispose();
+            }
+
+            public abstract AppendPrependIterator<TSource> Append(TSource element);
+
+            public abstract AppendPrependIterator<TSource> Prepend(TSource element);
+
+            public abstract TSource[] ToArray();
+
+            public abstract List<TSource> ToList();
+
+            public abstract int GetCount(bool onlyIfCheap);
+        }
+
+        private sealed class Append1Iterator<TSource> : AppendPrependIterator<TSource>
```

Comment:
I'm not following.  Why would it change the number of type checks or interface calls?  It would add branches, yes, because you'd need to check whether the bool was false or true to determine whether to append or prepend the element... why is it more than that?